External/FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp

```diff
@@ -67,6 +69,107 @@ extern char __start_HostToGuestTrampolineTemplate[];
 extern char __stop_HostToGuestTrampolineTemplate[];
 
 namespace FEXCore {
+  namespace VDSO {
+    using TimeType = decltype(::time)*;
+    using GetTimeOfDayType = decltype(::gettimeofday)*;
+    using ClockGetTimeType = decltype(::clock_gettime)*;
+    using ClockGetResType = decltype(::clock_getres)*;
+    using GetCPUType = decltype(::getcpu)*;
+
+    TimeType TimePtr = ::time;
+    GetTimeOfDayType GetTimeOfDayPtr = ::gettimeofday;
+    ClockGetTimeType ClockGetTimePtr = ::clock_gettime;
+    ClockGetResType ClockGetResPtr = ::clock_getres;
+    GetCPUType GetCPUPtr = ::getcpu;
+
+    static void time(void* ArgsRV) {
+      struct ArgsRV_t {
+        time_t *a_0;
+        uint64_t rv;
+      } *args = reinterpret_cast<ArgsRV_t*>(ArgsRV);
+
+      args->rv = TimePtr(args->a_0);
+    }
+
+    static void gettimeofday(void* ArgsRV) {
+      struct ArgsRV_t {
+        struct timeval *tv;
+        struct timezone *tz;
+        uint64_t rv;
+      } *args = reinterpret_cast<ArgsRV_t*>(ArgsRV);
+
+      args->rv = GetTimeOfDayPtr(args->tv, args->tz);
+    }
+
+    static void clock_gettime(void* ArgsRV) {
+      struct ArgsRV_t {
+        clockid_t clk_id;
+        struct timespec *tp;
+        uint64_t rv;
+      } *args = reinterpret_cast<ArgsRV_t*>(ArgsRV);
+
+      args->rv = ClockGetTimePtr(args->clk_id, args->tp);
+    }
+
+    static void clock_getres(void* ArgsRV) {
+      struct ArgsRV_t {
+        clockid_t clk_id;
+        struct timespec *tp;
+        uint64_t rv;
+      } *args = reinterpret_cast<ArgsRV_t*>(ArgsRV);
+
+      args->rv = ClockGetResPtr(args->clk_id, args->tp);
+    }
+
+    static void getcpu(void* ArgsRV) {
+      struct ArgsRV_t {
+        uint32_t *cpu;
+        uint32_t *node;
+        uint64_t rv;
+      } *args = reinterpret_cast<ArgsRV_t*>(ArgsRV);
+
+      args->rv = GetCPUPtr(args->cpu, args->node);
+    }
+
+    void LoadVDSO() {
+      void *vdso = dlopen("linux-vdso.so.1", RTLD_LAZY | RTLD_LOCAL | RTLD_NOLOAD);
```

Comment:
Hmm this is also strikingly similar to the contents of `ldr_ptrs.inl`, though the dlopen logic is slightly different.

I'm starting to understand why things are done this way: The `dlopen` logic here doesn't match `ldr_ptrs.inl`, hence the autogenerated unpacking functions don't work for us, hence the autogenerated export table doesn't work for us.

If I got this right, changing things so that Thunks.cpp can pass in a custom DL handle to a thunklib's InitFN would resolve this elegantly, wouldn't it? LoadVDSO would then do almost the same things as LoadLib, and all autogenerated entities could be used.

If this sounds good, I can help with the needed interface changes if you want.

Reply:
This `LoadVDSO` function has now been moved to the frontend as `LoadHostVDSO`.
I'm not sure if this would end up being worth making it in to the generic thunking interface. This is a one-off since there aren't any other thunks that are loaded in to userspace without filesystem accesses like VDSOs.

More ideally we would hardcode the guest side thunks directly in to the executable here, since binfmt_misc setups would then always have 2 pages loaded (one for 64-bit, one for 32-bit once that is in place). This would resolve the issue of VDSO thunks failing to be found inside a rootfs for example and not require an additional filesystem load. More matching what Linux on the host is doing.

But I didn't find a good way to have an `xxd` dependency, so a filesystem load and dependency it is.

Reply:
> This is a one-off since there aren't any other thunks that are loaded in to userspace without filesystem accesses like VDSOs.

The problem with "one-offs" in thunks is that they increase friction when making changes to the common thunk generator (particularly when autogenerated parts interact with handwritten code that mimics other autogenerated bits). There's a natural trade-off since some APIs (like vDSO, clearly) have special requirements, but my general guideline is to avoid hand-copying boilerplate that would otherwise be autogenerated.

In this case specifically, the manually implemented unpacking functions could come back to bite us once we implement efficient (i.e. pack-less) Guest<->Host transitions.

I'm fine to wave this PR through since it looks fine otherwise and there's no specific indicator this will cause problems in the future, particularly since it's just 5 actual functions anyway.

Reply:
Reasonable there. I wonder if in the future we could shift dependencies around to pull the generated unpacking code in to the frontend similar to this.

Reply:
> I wonder if in the future we could shift dependencies around to pull the generated unpacking code in to the frontend similar to this

Would this be just to avoid adding vDSO-specific interfaces to Thunks.cpp, or is there a deeper reason the unpacking code has to be compiled into the frontend rather than having an internal "libvdso-host.so" that we dlopen from the frontend?

Reply:
I guess some things here is that the loading is a quite a bit different than regular thunks.
- Guest library is only copied in to location, it can't have a static initializer to load the host side.
  - Need special handling for host load regardless.
- Host side loading doesn't load an equivalent library so it needs special casing
- There's an expectation that we will be missing symbols on the host side VDSO anyway, so we need special handling around pushing these to the libc implementation anyway. But we still want to try and fetch these from VDSO anyway since we may fix some of the missing symbols in the future. 
  - time & getcpu are likely candidates to get implementations if abused.
- Also another thunk implementation where missing host symbols aren't fatal because of the special handling.
  - We still need to provide the guest with an implementation regardless. Some applications treat these like a hard-dependency. (golang had a hard dependency for a while if I remember correctly?)